abi/abi.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package abi
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"strings"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/consts"
+
+	reflect "reflect"
+)
+
+type VM struct {
+	Actions []Action `serialize:"true" json:"actions"`
+}
+
+var _ codec.Typed = (*VM)(nil)
+
+const ABITypeID = consts.MaxUint8
+
+func (VM) GetTypeID() uint8 {
+	return ABITypeID
+}
+
+func (a *VM) Hash() [32]byte {
+	writer := codec.NewWriter(0, consts.NetworkSizeLimit)
+	err := codec.LinearCodec.MarshalInto(a, writer.Packer)
+	if err != nil {
+		// should never happen in prod, safe to panic
+		panic(fmt.Errorf("failed to marshal abi: MarshalInto: %w", err))
+	}
+	if writer.Err() != nil {
+		// should never happen in prod, safe to panic
+		panic(fmt.Errorf("failed to marshal abi: writer.Err: %w", writer.Err()))
+	}
+	abiHash := sha256.Sum256(writer.Bytes())
+	return abiHash
+}
+
+// Field represents a field in the VM (Application Binary Interface).
+type Field struct {
+	// Name of the field, overridden by the json tag if present
+	Name string `serialize:"true" json:"name"`
+	// Type of the field, either a Go type or struct name (excluding package name)
+	Type string `serialize:"true" json:"type"`
+}
+
+// Action represents the VM for an action.
+type Action struct {
+	ID    uint8  `serialize:"true" json:"id"`
+	Name  string `serialize:"true" json:"name"`
+	Types []Type `serialize:"true" json:"types"`
+}
+
+type Type struct {
+	Name   string  `serialize:"true" json:"name"`
+	Fields []Field `serialize:"true" json:"fields"`
+}
+
+func DescribeVM(actions []codec.Typed) (VM, error) {
```

Comment:
nit: rename to `New` (since callers will use `abi.New` to construct their abi).

Reply:
NewABI even better.

Reply:
Let's use `New` so that callers use `abi.New` instead of `abi.NewABI`